### ISSUE TITLE
IRIDA pipelines cannot handle Galaxy workflows with multiple of the same tool

### ISSUE THREAD

**Issue by apetkau** [phac-nml/irida#425]
## Describe the bug

The IRIDA pipeline execution cannot handle the case where there are multiple instances of the same tool in a Galaxy workflow.

## Steps to reproduce the problem

1. Create a workflow with multiple instances of the same tool in Galaxy.
2. Integrate into an IRIDA pipeline.
3. Try running the pipeline. You will get a similar error message about parameters:

```
08 Jul 2019 14:23:00,113 ERROR ca.corefacility.bioinformatics.irida.service.analysis.execution.AnalysisExecutionServiceAspect:65 - Error occured for submission: AnalysisSubmission [id=2, name=QITyping_20190708, submitter=admin, workflowId=4ad8ee74-251d-48b5-8b03-fa1dc656e8e7
, analysisState=SUBMITTING, analysisCleanedState=NOT_CLEANED] changing to state ERROR
java.lang.IllegalArgumentException: parameterMap already contains key with name "nucmer_min_id" mapping to value "90"
        at com.google.common.base.Preconditions.checkArgument(Preconditions.java:125)
        at ca.corefacility.bioinformatics.irida.service.analysis.workspace.galaxy.ParameterBuilderGalaxy.insertToMapRecursive(ParameterBuilderGalaxy.java:161)
        at ca.corefacility.bioinformatics.irida.service.analysis.workspace.galaxy.ParameterBuilderGalaxy.addParameter(ParameterBuilderGalaxy.java:132)                               
        at ca.corefacility.bioinformatics.irida.service.analysis.workspace.galaxy.AnalysisParameterServiceGalaxy.prepareAnalysisParameters(AnalysisParameterServiceGalaxy.java:86)
```

The specific line this is crashing at is:

https://github.com/phac-nml/irida/blob/2ae4f977c02b5099af50b7afec2b3083bb94043f/src/main/java/ca/corefacility/bioinformatics/irida/service/analysis/workspace/galaxy/ParameterBuilderGalaxy.java#L161

I am storing a map of tool parameter names to the Galaxy tool parameter name/values here. But, I am assuming that there is only one Galaxy tool with the same parameter name. If this does not hold, the above exception gets thrown.

## Expected behaviour

I expected IRIDA to be able to handle Galaxy workflows with multiple tools.

## Additional context

A screenshot of a workflow with multiple instances of the same tool (**ariba** but each instance varies the database).

![image](https://user-images.githubusercontent.com/200517/60839337-89c1fe00-a192-11e9-89ce-66dfcc0d1a00.png)

This is from this IRIDA pipeline <https://github.com/quadram-institute-bioscience/irida-plugin-qi-typing>.

@happykhan this is the source of the issue you are encountering. I will have to fix it up.

**Comment by apetkau** [comment]
Apparently I already have this an an issue at https://github.com/phac-nml/irida/issues/285. I'm closing that one since this one is more detailed.

**Comment by innovate-invent** [comment]
What is the status of this issue? It will prevent me from integrating any of my workflows.

**Comment by apetkau** [comment]
We've been discussing options to try and address this issue.

To get this to work we'd likely have to restructure IRIDA workflows to include the workflow step identifier for each tool in addition to the tool identifier:

```
        "1": {
            "tool_id": "toolshed.g2.bx.psu.edu/repos/nml/refseq_masher/refseq_masher_matches/0.1.1",
            "tool_version": "0.1.1",
```

So, here we'd have to include `1` as part of the information we record in IRIDA workflow XML file in addition to the tool_id e.g., in https://github.com/phac-nml/irida/blob/development/src/main/resources/ca/corefacility/bioinformatics/irida/model/workflow/analysis/type/workflows/RefSeqMasherOnPairedReads/0.1/irida_workflow.xml#L13

Then, we'd need to make use of blend4j to set an input parameter by the workflow identifier step along with the tool ID. This would be done in https://github.com/phac-nml/irida/blob/development/src/main/java/ca/corefacility/bioinformatics/irida/service/analysis/workspace/galaxy/AnalysisParameterServiceGalaxy.java#L94

It looks like this functionality exists already in blend4j https://github.com/galaxyproject/blend4j/blob/master/src/main/java/com/github/jmchilton/blend4j/galaxy/beans/WorkflowInputs.java#L61

So what we could do is just make an alternative field we define in the XML file, where instead of mapping a parameter to a `tool_id` we map to a workflow `step_id` (so, we replace `toolParameter` in https://github.com/phac-nml/irida/blob/development/src/main/resources/ca/corefacility/bioinformatics/irida/model/workflow/analysis/type/workflows/RefSeqMasherOnPairedReads/0.1/irida_workflow.xml#L13 with `workflowStepParameter`)

We would keep both options available so that previously-designed workflows still work, but for anything with duplicate tools people will have to use `workflowStepParameter` instead of `toolParameter` in the IRIDA workflow XML file to uniquely identify the parameters of each duplicate tool.

In other words, we create an instance of this class https://github.com/phac-nml/irida/blob/development/src/main/java/ca/corefacility/bioinformatics/irida/model/workflow/description/IridaToolParameter.java `IridaToolParameter`, but instead it's called `IridaWorkflowStepParameter`.

In a workflow, if such a parameter field exists, then in https://github.com/phac-nml/irida/blob/development/src/main/java/ca/corefacility/bioinformatics/irida/service/analysis/workspace/galaxy/AnalysisParameterServiceGalaxy.java#L94 we call `setStepParameter()` from blend4j (instead of `setToolParameter()`)

**Comment by apetkau** [comment]
So, a list of tasks to get this working would be:

* [ ] - Test out [setStepParameter()](https://github.com/galaxyproject/blend4j/blob/2201df98a54f8231f26b3594d14d920be2abfa25/src/main/java/com/github/jmchilton/blend4j/galaxy/beans/WorkflowInputs.java#L61) in `blend4j` to make sure it works as expected.
* [ ] - Implement a variant of [IridaToolParameter](https://github.com/phac-nml/irida/blob/639b0bbd621350d61a624279fcb6efc5fb7d29d9/src/main/java/ca/corefacility/bioinformatics/irida/model/workflow/description/IridaToolParameter.java) which allows you to map the workflow step id to a parameter value instead of the tool id.
* [ ] - Modify the Galaxy API call in https://github.com/phac-nml/irida/blob/639b0bbd621350d61a624279fcb6efc5fb7d29d9/src/main/java/ca/corefacility/bioinformatics/irida/service/analysis/workspace/galaxy/AnalysisParameterServiceGalaxy.java#L94 to set parameter values based on either the workflow step (or tool id)
* [ ] - Test out new code to verify it works (e.g., design a simple workflow with duplicate tools)
* [ ] - Work on updating [irida-wf-ga2xml](https://github.com/phac-nml/irida-wf-ga2xml) to support this functionality (though this can be done independent of any IRIDA release).

**Comment by innovate-invent** [comment]
I recommend targeting the uuid of the workflow inputs that specifically have types "data_input", "data_collection_input", and "parameter_input". This will provide the best forward compatibility with Galaxy. step_id is unstable and can change if the workflow is updated. Allowing tool inputs to be targeted directly bypasses Galaxies intent to declare all workflow inputs using the workflow input 'tools'.

```
"0": {
      "annotation": "",
      "content_id": null,
      "errors": null,
      "id": 0,
      "input_connections": {},
      "inputs": [],
      "label": "fasta",
      "name": "Input dataset collection",
      "outputs": [],
      "position": {
        "left": 254,
        "top": 474.33331298828125
      },
      "tool_id": null,
      "tool_state": "{\"collection_type\": \"list:list\"}",
      "tool_version": null,
      "type": "data_collection_input",
      "uuid": "7a81e28b-cab5-4cf8-a581-1044ee0b8516",
      "workflow_outputs": [
        {
          "label": null,
          "output_name": "output",
          "uuid": "8608591f-0d1a-42d0-86fc-5c3c97f51a07"
        }
      ]
    }
```

**Comment by apetkau** [comment]
@innovate-invent thanks for the suggestion. We will look into using the uuid of each step instead of the step id. I am uncertain if it will work directly with blend4j as it is right now, so we will have to test that out.

**Comment by innovate-invent** [comment]
You may have to do a initial lookup of the uuid to step id, but it will be worth it in the long run as you can just remove that step when blend4j catches up

**Comment by apetkau** [comment]
@innovate-invent I'm wondering, do you have an example of one of the workflows that you wanted to integrate but can't because of this issue?

**Comment by innovate-invent** [comment]
The one in mind hasn't been published yet. Do you need a real example or will a artificial one suffice?

[This is a simple example that includes two filter steps](https://github.com/phac-nml/irida/files/4587834/test_merge_empty.ga.tar.gz)

**Comment by apetkau** [comment]
Thanks, that helps me out.

**Comment by JeffreyThiessen** [comment]
As a workaround, you can use the `uuid` in the `toolId` field of the `irida_workflow.xml` to specify which tool the parameters are for.

Because of how blend4j works, both `toolid` and `stepid` get dumped into the same code block before sending to galaxy, so it doesn't matter which one you use, and `uuid` seems to work too, even though I couldn't find a spec for it.

**Comment by JeffreyThiessen** [comment]
This has been fixed with https://github.com/phac-nml/irida-wf-ga2xml/pull/12

Once we have a jar compiled and released this can be marked as complete and closed.

**Comment by apetkau** [comment]
This has been fixed with version **1.1.0** of `irida-wf-ga2xml` released at https://github.com/phac-nml/irida-wf-ga2xml/releases/tag/v1.1.0. This works by using the uuid in the IRIDA workflow XML file instead of the tool id by default.
